README.md
# floatsim

Approach to make a walkable/floatable simulation.

First approach to build a terrain with Perlin noise.
![First approach](screenshots/screenshot-20230308-1.png)

Then adding a cockpit (just a texture to start :)
![Add cockpit](screenshots/screenshot-20230308-2.png)

Tweaking the fog colors when changing depth. Adding a depth display.
![Depth and fog color](screenshots/screenshot-20230310-0.png)

Adding more HUD fragments: Depth meter and compass.
![Depth meter and compass](screenshots/screenshot-20230327-0.png)

Adding floating particles.
![Floating particles](screenshots/screenshot-20230402-0.png)

Bilinear terrain smoothin. Approach for a variometer. Loading ob OBJ files with materials.
![Bilinear terrain smoothing](screenshots/screenshot-20230411-0.png)
screenshot-20230411-0.png

## Coordinates

```
(height)
 Y
 |      Z (depth)
 |    /`
 |  /`
 |/`
 +----------X
        (width)
```

## Build THREE examples

```bash
npm run rollup-browser-umd
```

## Build for Testing

```bash
npm run compile-typescript
```

## Build for dist

CURRENTLY NOT WORKING, NEEDS TO BE FIXED

## Todos & ideas

- [DONE] Add basic scene setup.
- [DONE] Add simple terrain generation.
- Get class inheritance running.
- Get webpack back running (somehow the main.js script is falsy overwritten when packing).
- Build collision detection with terrain.
- Add dummy ship as simple composite object.
- [DONE] Build first person view.
- [DONE] Build controls (forward, backward, up, down, left, right).
- Add damping to the controls.
- [DONE] Add a non-terrain object (a building or so). Added an OBJ concrete cylinder.
- Build collision detection with non-terrain objects.
- [DONE] Add FPS stats.
- Limit upper depth at 0 (or less). Vessel must never go over zero (which would be above the sea level).
- Build a caustics animation/shader for the ocean floor.
- [DONE] Change water color depending of depth.
- [DONE] Add cockpit (image).
- Add cockpit (mesh?).
- Make fancy scene blending effect (like in Archimedean Dynasty cutscenes).
- [DONE] Add floating particles (sprites?).
- [DONE] Add ambient sound (like water bubbling/streaming).
- Add collision sounds. Each "material" should have its own sound (sand, metal, stone, organic?).
- [DONE] Add a depth measure (to the ground and to the surface). Add a "pressure" HUD.
- Add a system alarm to warn for "radiation" and "high pressure".
- [DONE] Add add compass HUD. (use OrbitControls.getAzimuthalAngle() ?)
- Add water plants.
- Add a sonar HUD.
- [DONE] Let initial settings being passed by GET parameters.
- Tinker with THREE VolumetricLight.
- Fog Color must be passed to the new terrain shader on each position change.
- Add a custom camera/scene for separate cockpit rendering.
- Create simple logo and favicon?
- Add target HUD like CubeIconBox (https://hofk.de/main/discourse.threejs/2019/CubeBoxIcon/CubeBoxIcon.html)
- Make hudData a globally available object?
- HudComponent: buffer color style string in class variable (is rarely changed)
- Compass: load compass texture from SVG?
- Add floating particles (sprites?).
- Bounds2Immutable: try function overloading with the new Typescript here (constructor).
- Add terrain bounds + respawn for particles.
- Add variometer HUD for displaying current ship angle (up/down).
- Pick a nice font to use in all HUDs.
- Bring the variometer to correct display (measure up/down angle of ship).
- make numToChar available for more than 26 numbers.

## Credits

- Ken Perlin for the improved Perlin noise https://cs.nyu.edu/~perlin/noise/
- jaamo for the HUD howto (https://codepen.io/sjcobb/pen/XWjMdXb)
- Three JS Custom Shader Material and fog hack: https://snayss.medium.com/three-js-fog-hacks-fc0b42f63386
- Underwater ambience sound https://pixabay.com/sound-effects/underwater-ambiencewav-14428/
- Cocopon for Tweakpane https://cocopon.github.io/tweakpane
- Vector Ruler Generator by robbbb https://robbbb.github.io/VectorRuler/
- pailhead011 for the tutorial of how to extend existing THREEJS shaders and pass custom uniforms to them https://medium.com/@pailhead011/extending-three-js-materials-with-glsl-78ea7bbb9270
- Sam Hocevar for the rgb2hsl and reverse GLSL functions https://gamedev.stackexchange.com/questions/59797/glsl-shader-change-hue-saturation-brightness


changelog.md
- 2023-10-20
  - Adding nice icons for battery and thermometer.
- 2023-10-19
  - Adding plotboilerplate@1.21.0
  - Tweaking hud positions calculation (seems still kinda wrong, though).
- 2023-09-24
  - Adding proper modulalization for the typewriter effect.
  - Adding a MediaStorage to avoid multiple media request.
  - Added a quick logo.
- 2023-09-22
  - Adding hooks to build a typewriter effect.
  - Adding intro overlay with typewriter effect.
  - Adding intro with map preview.
  - Adding initial shutter effect.
- 2023-09-09
  - Adding preact to render proper JSX. This is intended to render some frontend ui stuff.
  - Getting preact running in as isloated container.
  - Refactoring the overlay to work completely with preact now.
  - Adding a listener logic to the SceneContainer, to listen for gameReady and gameRunning/gamePaused events.
- 2023-09-08
  - Adding a dialigue library (rpg-dialigue-js).
- 2023-05-03
  - Refactored the caustic shader to a new version to make it usable with existing threejs shaders.
- 2023-04-29
  - Added HUD param: groundDepth.
- 2023-04-26
  - Chaninging HudData.shiprotation from {y,y,z} to {upAngle}.
  - Adding virtual horizon to HUD.
- 2023-04-25
  - Adding simple buoy meshes and navpoint rendering.
- 2023-04-22
  - Refactored the Caustic shader library (updating exports).
  - Refactoring test cutscene shader to shader library in the simulation.
  - Adding a Navpoints HUD fragment.
- 2023-04-20
  - Added a cutscene shader test.
- 2023-03-xx
  - Updating Perlin Terrain: int to float and added smoothening.
  - Added ambient sound.
  - Added HUD fragment: Variometer.
  - Added HUD fragment: Depthmeter.
  - Added HUD component: Compass.
  - Added HUD fragment: general info display.
  - Added Perlin terrain generation (test).
  - Added caustic shader.
  - Added fog.


src/ts/utils/perlin.ts
/*
 * A speed-improved perlin and simplex noise algorithms for 2D.
 *
 * Based on example code by Stefan Gustavson (stegu@itn.liu.se).
 * Optimisations by Peter Eastman (peastman@drizzle.stanford.edu).
 * Better rank ordering method by Stefan Gustavson in 2012.
 * Converted to Javascript by Joseph Gentle.
 *
 * Version 2012-03-09
 *
 * This code was placed in the public domain by its original author,
 * Stefan Gustavson. You may use it as you see fit, but
 * attribution is appreciated.
 *
 * @modified 2023-03-07 Ikaros Kappler; ported to Typescript & refactored to class.
 */

class Grad {
  x: number;
  y: number;
  z: number;
  constructor(x: number, y: number, z: number) {
    this.x = x;
    this.y = y;
    this.z = z;
  }
  dot2(x: number, y: number) {
    return this.x * x + this.y * y;
  }
  dot3(x: number, y: number, z: number) {
    return this.x * x + this.y * y + this.z * z;
  }
}

const grad3: Array<Grad> = [
  new Grad(1, 1, 0),
  new Grad(-1, 1, 0),
  new Grad(1, -1, 0),
  new Grad(-1, -1, 0),
  new Grad(1, 0, 1),
  new Grad(-1, 0, 1),
  new Grad(1, 0, -1),
  new Grad(-1, 0, -1),
  new Grad(0, 1, 1),
  new Grad(0, -1, 1),
  new Grad(0, 1, -1),
  new Grad(0, -1, -1)
];

const p: Array<number> = [
  151, 160, 137, 91, 90, 15, 131, 13, 201, 95, 96, 53, 194, 233, 7, 225, 140, 36, 103, 30, 69, 142, 8, 99, 37, 240, 21, 10, 23,
  190, 6, 148, 247, 120, 234, 75, 0, 26, 197, 62, 94, 252, 219, 203, 117, 35, 11, 32, 57, 177, 33, 88, 237, 149, 56, 87, 174, 20,
  125, 136, 171, 168, 68, 175, 74, 165, 71, 134, 139, 48, 27, 166, 77, 146, 158, 231, 83, 111, 229, 122, 60, 211, 133, 230, 220,
  105, 92, 41, 55, 46, 245, 40, 244, 102, 143, 54, 65, 25, 63, 161, 1, 216, 80, 73, 209, 76, 132, 187, 208, 89, 18, 169, 200, 196,
  135, 130, 116, 188, 159, 86, 164, 100, 109, 198, 173, 186, 3, 64, 52, 217, 226, 250, 124, 123, 5, 202, 38, 147, 118, 126, 255,
  82, 85, 212, 207, 206, 59, 227, 47, 16, 58, 17, 182, 189, 28, 42, 223, 183, 170, 213, 119, 248, 152, 2, 44, 154, 163, 70, 221,
  153, 101, 155, 167, 43, 172, 9, 129, 22, 39, 253, 19, 98, 108, 110, 79, 113, 224, 232, 178, 185, 112, 104, 218, 246, 97, 228,
  251, 34, 242, 193, 238, 210, 144, 12, 191, 179, 162, 241, 81, 51, 145, 235, 249, 14, 239, 107, 49, 192, 214, 31, 181, 199, 106,
  157, 184, 84, 204, 176, 115, 121, 50, 45, 127, 4, 150, 254, 138, 236, 205, 93, 222, 114, 67, 29, 24, 72, 243, 141, 128, 195, 78,
  66, 215, 61, 156, 180
];
// To remove the need for index wrapping, double the permutation table length
const perm = new Array<number>(512);
const gradP = new Array<Grad>(512);

// Skewing and unskewing factors for 2, 3, and 4 dimensions
const F2 = 0.5 * (Math.sqrt(3) - 1);
const G2 = (3 - Math.sqrt(3)) / 6;

const F3 = 1 / 3;
const G3 = 1 / 6;

const fade = (t: number) => {
  return t * t * t * (t * (t * 6 - 15) + 10);
};

const lerp = (a: number, b: number, t: number) => {
  return (1 - t) * a + t * b;
};

export const noise = {
  // This isn't a very good seeding function, but it works ok. It supports 2^16
  // different seed values. Write something better if you need more seeds.
  //   module.seed = function (seed) {
  seed: (seed: number) => {
    if (seed > 0 && seed < 1) {
      // Scale the seed out
      seed *= 65536;
    }

    seed = Math.floor(seed);
    if (seed < 256) {
      seed |= seed << 8;
    }

    for (var i = 0; i < 256; i++) {
      var v: number;
      if (i & 1) {
        v = p[i] ^ (seed & 255);
      } else {
        v = p[i] ^ ((seed >> 8) & 255);
      }

      perm[i] = perm[i + 256] = v;
      gradP[i] = gradP[i + 256] = grad3[v % 12];
    }
  },

  // 2D simplex noise
  simplex2: (xin: number, yin: number) => {
    var n0: number, n1: number, n2: number; // Noise contributions from the three corners
    // Skew the input space to determine which simplex cell we're in
    const s = (xin + yin) * F2; // Hairy factor for 2D
    var i: number = Math.floor(xin + s);
    var j: number = Math.floor(yin + s);
    const t: number = (i + j) * G2;
    const x0: number = xin - i + t; // The x,y distances from the cell origin, unskewed.
    const y0: number = yin - j + t;
    // For the 2D case, the simplex shape is an equilateral triangle.
    // Determine which simplex we are in.
    var i1: number, j1: number; // Offsets for second (middle) corner of simplex in (i,j) coords
    if (x0 > y0) {
      // lower triangle, XY order: (0,0)->(1,0)->(1,1)
      i1 = 1;
      j1 = 0;
    } else {
      // upper triangle, YX order: (0,0)->(0,1)->(1,1)
      i1 = 0;
      j1 = 1;
    }
    // A step of (1,0) in (i,j) means a step of (1-c,-c) in (x,y), and
    // a step of (0,1) in (i,j) means a step of (-c,1-c) in (x,y), where
    // c = (3-sqrt(3))/6
    const x1 = x0 - i1 + G2; // Offsets for middle corner in (x,y) unskewed coords
    const y1 = y0 - j1 + G2;
    const x2 = x0 - 1 + 2 * G2; // Offsets for last corner in (x,y) unskewed coords
    const y2 = y0 - 1 + 2 * G2;
    // Work out the hashed gradient indices of the three simplex corners
    i &= 255;
    j &= 255;
    const gi0 = gradP[i + perm[j]];
    const gi1 = gradP[i + i1 + perm[j + j1]];
    const gi2 = gradP[i + 1 + perm[j + 1]];
    // Calculate the contribution from the three corners
    var t0 = 0.5 - x0 * x0 - y0 * y0;
    if (t0 < 0) {
      n0 = 0;
    } else {
      t0 *= t0;
      n0 = t0 * t0 * gi0.dot2(x0, y0); // (x,y) of grad3 used for 2D gradient
    }
    var t1 = 0.5 - x1 * x1 - y1 * y1;
    if (t1 < 0) {
      n1 = 0;
    } else {
      t1 *= t1;
      n1 = t1 * t1 * gi1.dot2(x1, y1);
    }
    var t2 = 0.5 - x2 * x2 - y2 * y2;
    if (t2 < 0) {
      n2 = 0;
    } else {
      t2 *= t2;
      n2 = t2 * t2 * gi2.dot2(x2, y2);
    }
    // Add contributions from each corner to get the final noise value.
    // The result is scaled to return values in the interval [-1,1].
    return 70 * (n0 + n1 + n2);
  },

  // 3D simplex noise
  simplex3: (xin: number, yin: number, zin: number) => {
    var n0: number, n1: number, n2: number, n3: number; // Noise contributions from the four corners

    // Skew the input space to determine which simplex cell we're in
    const s: number = (xin + yin + zin) * F3; // Hairy factor for 2D
    var i: number = Math.floor(xin + s);
    var j: number = Math.floor(yin + s);
    var k: number = Math.floor(zin + s);

    const t: number = (i + j + k) * G3;
    const x0: number = xin - i + t; // The x,y distances from the cell origin, unskewed.
    const y0: number = yin - j + t;
    const z0: number = zin - k + t;

    // For the 3D case, the simplex shape is a slightly irregular tetrahedron.
    // Determine which simplex we are in.
    var i1: number, j1: number, k1: number; // Offsets for second corner of simplex in (i,j,k) coords
    var i2: number, j2: number, k2: number; // Offsets for third corner of simplex in (i,j,k) coords
    if (x0 >= y0) {
      if (y0 >= z0) {
        i1 = 1;
        j1 = 0;
        k1 = 0;
        i2 = 1;
        j2 = 1;
        k2 = 0;
      } else if (x0 >= z0) {
        i1 = 1;
        j1 = 0;
        k1 = 0;
        i2 = 1;
        j2 = 0;
        k2 = 1;
      } else {
        i1 = 0;
        j1 = 0;
        k1 = 1;
        i2 = 1;
        j2 = 0;
        k2 = 1;
      }
    } else {
      if (y0 < z0) {
        i1 = 0;
        j1 = 0;
        k1 = 1;
        i2 = 0;
        j2 = 1;
        k2 = 1;
      } else if (x0 < z0) {
        i1 = 0;
        j1 = 1;
        k1 = 0;
        i2 = 0;
        j2 = 1;
        k2 = 1;
      } else {
        i1 = 0;
        j1 = 1;
        k1 = 0;
        i2 = 1;
        j2 = 1;
        k2 = 0;
      }
    }
    // A step of (1,0,0) in (i,j,k) means a step of (1-c,-c,-c) in (x,y,z),
    // a step of (0,1,0) in (i,j,k) means a step of (-c,1-c,-c) in (x,y,z), and
    // a step of (0,0,1) in (i,j,k) means a step of (-c,-c,1-c) in (x,y,z), where
    // c = 1/6.
    const x1: number = x0 - i1 + G3; // Offsets for second corner
    const y1: number = y0 - j1 + G3;
    const z1: number = z0 - k1 + G3;

    const x2: number = x0 - i2 + 2 * G3; // Offsets for third corner
    const y2: number = y0 - j2 + 2 * G3;
    const z2: number = z0 - k2 + 2 * G3;

    const x3: number = x0 - 1 + 3 * G3; // Offsets for fourth corner
    const y3: number = y0 - 1 + 3 * G3;
    const z3: number = z0 - 1 + 3 * G3;

    // Work out the hashed gradient indices of the four simplex corners
    i &= 255;
    j &= 255;
    k &= 255;
    const gi0: Grad = gradP[i + perm[j + perm[k]]];
    const gi1: Grad = gradP[i + i1 + perm[j + j1 + perm[k + k1]]];
    const gi2: Grad = gradP[i + i2 + perm[j + j2 + perm[k + k2]]];
    const gi3: Grad = gradP[i + 1 + perm[j + 1 + perm[k + 1]]];

    // Calculate the contribution from the four corners
    var t0: number = 0.5 - x0 * x0 - y0 * y0 - z0 * z0;
    if (t0 < 0) {
      n0 = 0;
    } else {
      t0 *= t0;
      n0 = t0 * t0 * gi0.dot3(x0, y0, z0); // (x,y) of grad3 used for 2D gradient
    }
    var t1 = 0.5 - x1 * x1 - y1 * y1 - z1 * z1;
    if (t1 < 0) {
      n1 = 0;
    } else {
      t1 *= t1;
      n1 = t1 * t1 * gi1.dot3(x1, y1, z1);
    }
    var t2 = 0.5 - x2 * x2 - y2 * y2 - z2 * z2;
    if (t2 < 0) {
      n2 = 0;
    } else {
      t2 *= t2;
      n2 = t2 * t2 * gi2.dot3(x2, y2, z2);
    }
    var t3 = 0.5 - x3 * x3 - y3 * y3 - z3 * z3;
    if (t3 < 0) {
      n3 = 0;
    } else {
      t3 *= t3;
      n3 = t3 * t3 * gi3.dot3(x3, y3, z3);
    }
    // Add contributions from each corner to get the final noise value.
    // The result is scaled to return values in the interval [-1,1].
    return 32 * (n0 + n1 + n2 + n3);
  },

  // 2D Perlin Noise
  perlin2: function (x_init: number, y_init: number) {
    // Find unit grid cell containing point
    var X: number = Math.floor(x_init),
      Y: number = Math.floor(y_init);
    // Get relative xy coordinates of point within that cell
    const x: number = x_init - X;
    const y: number = y_init - Y;
    // Wrap the integer cells at 255 (smaller integer period can be introduced here)
    X = X & 255;
    Y = Y & 255;

    // Calculate noise contributions from each of the four corners
    const n00: number = gradP[X + perm[Y]].dot2(x, y);
    const n01: number = gradP[X + perm[Y + 1]].dot2(x, y - 1);
    const n10: number = gradP[X + 1 + perm[Y]].dot2(x - 1, y);
    const n11: number = gradP[X + 1 + perm[Y + 1]].dot2(x - 1, y - 1);

    // Compute the fade curve value for x
    const u: number = fade(x);

    // Interpolate the four results
    return lerp(lerp(n00, n10, u), lerp(n01, n11, u), fade(y));
  },

  // 3D Perlin Noise
  perlin3: (x_init: number, y_init: number, z_init: number) => {
    // Find unit grid cell containing point
    var X: number = Math.floor(x_init),
      Y: number = Math.floor(y_init),
      Z: number = Math.floor(z_init);
    // Get relative xyz coordinates of point within that cell
    const x: number = x_init - X;
    const y: number = y_init - Y;
    const z: number = z_init - Z;
    // Wrap the integer cells at 255 (smaller integer period can be introduced here)
    X = X & 255;
    Y = Y & 255;
    Z = Z & 255;

    // Calculate noise contributions from each of the eight corners
    const n000: number = gradP[X + perm[Y + perm[Z]]].dot3(x, y, z);
    const n001: number = gradP[X + perm[Y + perm[Z + 1]]].dot3(x, y, z - 1);
    const n010: number = gradP[X + perm[Y + 1 + perm[Z]]].dot3(x, y - 1, z);
    const n011: number = gradP[X + perm[Y + 1 + perm[Z + 1]]].dot3(x, y - 1, z - 1);
    const n100: number = gradP[X + 1 + perm[Y + perm[Z]]].dot3(x - 1, y, z);
    const n101: number = gradP[X + 1 + perm[Y + perm[Z + 1]]].dot3(x - 1, y, z - 1);
    const n110: number = gradP[X + 1 + perm[Y + 1 + perm[Z]]].dot3(x - 1, y - 1, z);
    const n111: number = gradP[X + 1 + perm[Y + 1 + perm[Z + 1]]].dot3(x - 1, y - 1, z - 1);

    // Compute the fade curve value for x, y, z
    const u: number = fade(x);
    const v: number = fade(y);
    const w: number = fade(z);

    // Interpolate
    return lerp(lerp(lerp(n000, n100, u), lerp(n001, n101, u), w), lerp(lerp(n010, n110, u), lerp(n011, n111, u), w), v);
  }
};

noise.seed(0);
